blog/2024-02-05-compressed-fsm.md

```diff
@@ -0,0 +1,138 @@
+, it always prefers to combine it with a following 
+<code style="color: black; background-color: lightblue;">,</code>
+to form a more frequent token
+<code style="color: black; background-color: lightblue;">",</code>
+. This effect may cause some strange behaviors. For example, in the above case, if the regex is set to
+<code style="color: black; background-color: lightblue;">"[\w\d\s]*"</code>
+(without the last 
+<code style="color: black; background-color: lightblue;">,</code>
+), it can lead to endless decoding because an LLM wants to stop with <code style="color: black; background-color: lightblue;">",</code> but this token is not allowed.
+
+Moreover, during jump-forward decoding, we've found that different tokenization strategies to the jump-forwarded part may lead to different logit distributions for the subsequent tokens. Simply appending the tokenized jump-forwarded section to the current token sequence might yield unexpected outcomes.
+
+To manage these issues, we propose the following solutions:
+- Prefer the use of a comprehensive regular expression to guide the entire decoding process, rather than employing multiple concatenated regular expressions. This approach ensures that both Finite State Machines (FSM) and Large Language Models (LLM) are cognizant of the entire decoding process, thereby minimizing boundary-related issues as much as possible.
+- We have implemented a re-tokenization mechanism during the jump-forward phase. This involves appending the string instead of the tokens, followed by a re-tokenization of the entire text. This method effectively resolves most tokenization issues and results in only a minor increase in computational overhead, approximately 4\%.
+
+## Benchmark Results
+
+We benchmarked our jump-forward decoding on two tasks:
+
+- Crafting a character's data in JSON format, guided by a brief prompt.
+- Extracting a city's information from a long document and outputing it in JSON format.
+
+We tested llama-7B on an NVIDIA A10 GPU (24GB), and used vllm v0.2.7, guidance v0.1.0, outlines v0.2.5 and llama.cpp v0.2.38(Python binding) . The figure below shows the throughput (using the maximum batch size supported by each system) and latency (with a batch size of 1) of these methods:
+
+<img src="/images/blog/compressed_fsm/result.png" style="width: 100%; max-width: 100%; margin-left: auto; margin-right: auto; margin-bottom: auto"></img>
+<p style="color:gray; text-align: center;">
+Figure 6: Benchmark Results
+</p>
+
+The results show that SGLang with our decoding algorithm significantly outperforms all other systems.
+It can reduce the latency by up to 2x and boost throughput by up to 2.5x.
+We suspect that the higher throughput of SGLang with no Jump-Forward compared to Outlines+vLLM is due to some overhead in Outlines.
+
+We have been testing this feature with [Boson.ai](https://boson.ai/) for two weeks, who are bringing this feature into their production use cases due to its effectiveness.
```

Comment:
```suggestion
We have been testing this feature with [Boson.ai](https://boson.ai/) for two weeks, who are bringing this feature into their production use cases because it guarantees robust response with higher decoding throughput.
```